### PR TITLE
[SPARK-37383][SQL][WEBUI]Show the parsing time for each phase of a SQL on spark ui

### PR DESCRIPTION
**Pull request by caican00**
### What changes were proposed in this pull request?
The PR is proposed to:
show the timeSpent for each phase of a SQL on spark ui.
image

image

The implications of the following indicators
```
parsing:   parser phase
analysis:   analysis phase
optimization:   optimization phase
planning:   physical plan phase
```

### Why are the changes needed?
the time spent for each parsing phase of a SQL is counted and recorded in QueryPlanningTracker , but it is not yet shown anywhere.
when SQL parsing is suspected to be slow, we cannot confirm which phase is slow，therefore, it is necessary to show the SQL parsing time.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
modified tests.

### CODE REVIEW

**Comment by caican00** [comment]
@cloud-fan Could you please help review the code? thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 1]
```diff
@@ -120,6 +120,24 @@ class QueryPlanningTracker {
     ret
   }
 
+  /**
+   * print out the timeSpent for each phase of a SQL
+   */
+  def logTimeSpent(): Unit = {
+    var totalTimeSpent = 0L
+    val timeSpentSummary: StringBuffer = new StringBuffer()
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:128]: Maybe `StringBuilder ` is ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 2]
```diff
@@ -19,9 +19,9 @@ package org.apache.spark.sql.catalyst
 
 import scala.collection.JavaConverters._
 
+import org.apache.spark.internal.Logging
 import org.apache.spark.util.BoundedPriorityQueue
 
-
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:24]: no need to remove this line

Comment (caican00) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:24]: thanks! updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala (hunk 1 above)
Comment (caican00) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: It seems so. thanks, updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala (hunk 2 above)
Comment (caican00) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:24]: thanks, updated

**Comment by caican00** [comment]
cc @rxin

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala [hunk 3]
```diff
@@ -90,7 +91,7 @@ object QueryPlanningTracker {
 }
 
 
-class QueryPlanningTracker {
+class QueryPlanningTracker extends Logging {
```
[the remark this answers is not in the document]
Comment (caican00) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala:94]: It seems like a good idea, i'll try to fix it

**Comment by LuciferYang** [comment]
@caican00 You should add a screenshot to the pr description and highlight the specific changes because this pr will cause `WEBUI` changes.

**Comment by caican00** [comment]
```
@caican00 You should add a screenshot to the pr description and highlight the specific changes because this pr will cause WEBUI changes.

Need add [WEBUI] to pr title
```
@LuciferYang Thank you for your suggestion so much. I have updated it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (caican00) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/QueryPlanningTracker.scala]: @cloud-fan Hi, could you review this patch again? thanks

**Comment by caican00** [comment]
> @caican00 You should add a screenshot to the pr description and highlight the specific changes because this pr will cause `WEBUI` changes.
> 
> Need add `[WEBUI]` to pr title

@LuciferYang Hi, i have update it. Could you help to review this patch again? thanks

**Comment by yliou** [comment]
Hi @caican00, I was working on the same idea but have a different approach at https://github.com/apache/spark/pull/35939 which also has a REST endpoint and Spark rule timing information.

**Pull request closed without merging**